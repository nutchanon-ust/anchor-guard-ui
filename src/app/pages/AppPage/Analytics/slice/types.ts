/* --- STATE --- */
export interface AnalyticsState {
  bidPools: BidPool[];
  liquidationProfile: LiquidationProfile[];
  lunaPrice: number;
}

export interface BidPool {
  premium: number;
  totalBidAmount: number;
  type?: string;
}

export interface LiquidationProfile {
  Date: string;
  Luna_Liquidation_Price: number;
  Loan_Value: number;
}
