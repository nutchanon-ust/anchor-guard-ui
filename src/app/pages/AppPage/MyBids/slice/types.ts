import { Repo } from 'types/Repo';

/* --- STATE --- */
export interface MyBidsState {
  myBids: Bid[];
  loading: boolean;
  walletAddress: string | null;
}

export interface Bid {
  id: string;
  premium: string;
  bidRemaining: string;
  bidStatus: string;
  amountFilled: string;
  waitEnd: number;
  timeLeft: number;
}

/* 
  If you want to use 'ContainerState' keyword everywhere in your feature folder, 
  instead of the 'HomePageState' keyword.
*/
export type ContainerState = MyBidsState;
