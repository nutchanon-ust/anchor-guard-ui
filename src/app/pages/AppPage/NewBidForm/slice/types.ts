/* --- STATE --- */
export interface NewBidState {
  userUstBalance: number;
}
