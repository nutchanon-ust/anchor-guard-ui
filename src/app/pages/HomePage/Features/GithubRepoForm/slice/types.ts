import { Repo } from 'types/Repo';

/* --- STATE --- */
export interface GithubRepoFormState {
  username: string;
  loading: boolean;
  error?: RepoErrorType | null;
  repositories: Repo[];
}

export enum RepoErrorType {
  RESPONSE_ERROR = 1,
  USER_NOT_FOUND = 2,
  USERNAME_EMPTY = 3,
  USER_HAS_NO_REPO = 4,
  GITHUB_RATE_LIMIT = 5,
}

/* 
  If you want to use 'ContainerState' keyword everywhere in your feature folder, 
  instead of the 'HomePageState' keyword.
*/
export type ContainerState = GithubRepoFormState;
