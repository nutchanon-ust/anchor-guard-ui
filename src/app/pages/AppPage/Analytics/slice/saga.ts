import { LCDClient } from '@terra-money/terra.js';
import { NetworkInfo } from '@terra-money/wallet-provider';
import {
  selectLCDClient,
  selectNetwork,
} from 'app/components/WalletSelector/slice/selectors';
import {
  ANCHOR_LIQUIDATION_QUEUE_CONTRACT_ADDRESS,
  BLUNA_MAINNET_ADDRESS,
} from 'app/constants';
import { formatUnits } from 'ethers/lib/utils';
import { take, call, put, select, takeLatest } from 'redux-saga/effects';
import { request } from 'utils/request';
import { analyticsActions as actions } from '.';
import { selectWalletAddress } from '../../MyBids/slice/selectors';
import { selectCollateralToken } from '../../NewBidForm/slice/selectors';
import { LiquidationProfile } from './types';

// function* doSomething() {}
export function* getLiquidationProfile() {
  try {
    const liq: LiquidationProfile[] = yield call(
      request,
      'https://api.alphadefi.fund/historical/kujira/profile',
    );
    yield put(actions.liquidationProfileLoaded(liq));
  } catch (e) {
    yield put(actions.liquidationProfileLoaded([]));
  }
}

export function* getBidPool() {
  const lcd: LCDClient = yield select(selectLCDClient);
  const collateralToken: string =
    (yield select(selectCollateralToken)) || BLUNA_MAINNET_ADDRESS;
  const network: NetworkInfo = yield select(selectNetwork);
  if (collateralToken) {
    const result = yield Promise.all(
      Array.from(Array(31).keys()).map(slot => {
        return lcd.wasm.contractQuery(
          ANCHOR_LIQUIDATION_QUEUE_CONTRACT_ADDRESS(network),
          {
            bid_pool: {
              collateral_token: collateralToken,
              bid_slot: slot,
            },
          },
        );
      }),
    );
    yield put(
      actions.loaded(
        result.map((each, index) => {
          return {
            premium: index,
            totalBidAmount: each.total_bid_amount / 1e6,
          };
        }),
      ),
    );
  }
}

export function* analyticsSaga() {
  yield takeLatest(actions.getBidPool.type, getBidPool);
  yield takeLatest(actions.getLiquidationProfile.type, getLiquidationProfile);
}
