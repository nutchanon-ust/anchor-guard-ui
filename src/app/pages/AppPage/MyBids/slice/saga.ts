import { call, put, select, takeLatest, delay } from 'redux-saga/effects';
import { myBidsActions as actions } from '.';
import { LCDClient } from '@terra-money/terra.js';
import { formatUnits } from 'ethers/lib/utils';
import {
  ANCHOR_LIQUIDATION_QUEUE_CONTRACT_ADDRESS,
  ANCHOR_LIQUIDATION_QUEUE_TESTNET_CONTRACT_ADDRESS,
  BLUNA_TESTNET_ADDRESS,
  TESTNET_CHAIN_ID,
  TESTNET_LCD_URL,
} from 'app/constants';
import { selectWalletAddress } from './selectors';

export function* getMyBids() {
  console.log('terra');
  const terra = new LCDClient({
    URL: TESTNET_LCD_URL,
    chainID: TESTNET_CHAIN_ID,
  });

  const walletAddress: string = yield select(selectWalletAddress);
  if (walletAddress) {
    const result = yield terra.wasm.contractQuery(
      ANCHOR_LIQUIDATION_QUEUE_TESTNET_CONTRACT_ADDRESS,
      {
        bids_by_user: {
          collateral_token: BLUNA_TESTNET_ADDRESS,
          bidder: walletAddress,
        },
      }, // query msg
    );
    if (result) {
      console.log(result);
      yield put(
        actions.loaded(
          result.bids.map(each => ({
            id: each.idx,
            premium: each.premium_slot,
            bidRemaining: Number(formatUnits(each.amount, 6)),
            bidStatus: each.wait_end === null ? 'Active' : 'Pending',
            amountFilled: Number(
              formatUnits(each.pending_liquidated_collateral, 6),
            ),
            waitEnd: each.wait_end,
          })),
        ),
      );
    } else {
      yield put(actions.loaded([]));
    }
  } else {
    yield put(actions.loaded([]));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export function* myBidsSaga() {
  // Watches for loadRepos actions and calls getRepos when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield takeLatest(actions.load.type, getMyBids);
}
