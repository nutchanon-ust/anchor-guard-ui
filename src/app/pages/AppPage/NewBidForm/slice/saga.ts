// import { take, call, put, select, takeLatest } from 'redux-saga/effects';
// import { newBidActions as actions } from '.';

// function* doSomething() {}

export function* newBidSaga() {
  // yield takeLatest(actions.someAction.type, doSomething);
}
