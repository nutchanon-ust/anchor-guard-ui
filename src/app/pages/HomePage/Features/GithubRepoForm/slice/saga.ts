import { call, put, select, takeLatest, delay } from 'redux-saga/effects';
import { request } from 'utils/request';
import { selectUsername } from './selectors';
import { githubRepoFormActions as actions } from '.';
import { Repo } from 'types/Repo';
import { RepoErrorType } from './types';

/**
 * Github repos request/response handler
 */
export function* getRepos() {
  yield delay(500);
  // Select username from store
  const username: string = yield select(selectUsername);
  if (username.length === 0) {
    yield put(actions.repoError(RepoErrorType.USERNAME_EMPTY));
    return;
  }
  const requestURL = `https://api.github.com/users/${username}/repos?type=all&sort=updated`;

  try {
    // Call our request helper (see 'utils/request')
    const repos: Repo[] = yield call(request, requestURL);
    if (repos?.length > 0) {
      yield put(actions.reposLoaded(repos));
    } else {
      yield put(actions.repoError(RepoErrorType.USER_HAS_NO_REPO));
    }
  } catch (err: any) {
    if (err.response?.status === 404) {
      yield put(actions.repoError(RepoErrorType.USER_NOT_FOUND));
    } else if (err.message === 'Failed to fetch') {
      yield put(actions.repoError(RepoErrorType.GITHUB_RATE_LIMIT));
    } else {
      yield put(actions.repoError(RepoErrorType.RESPONSE_ERROR));
    }
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export function* githubRepoFormSaga() {
  // Watches for loadRepos actions and calls getRepos when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield takeLatest(actions.loadRepos.type, getRepos);
}
