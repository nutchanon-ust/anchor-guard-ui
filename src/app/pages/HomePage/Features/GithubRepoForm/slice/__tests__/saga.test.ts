import { put, takeLatest } from 'redux-saga/effects';
import * as slice from '..';

import { githubRepoFormSaga, getRepos } from '../saga';
import { RepoErrorType } from '../types';

describe('getRepos Saga', () => {
  let username: any;
  let repos: any;
  let getReposIterator: ReturnType<typeof getRepos>;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getReposIterator = getRepos();
    const delayDescriptor = getReposIterator.next().value;
    expect(delayDescriptor).toMatchSnapshot();

    const selectDescriptor = getReposIterator.next().value;
    expect(selectDescriptor).toMatchSnapshot();
  });

  it('should return error if username is empty', () => {
    username = '';
    const putDescriptor = getReposIterator.next(username).value;
    expect(putDescriptor).toEqual(
      put(slice.githubRepoFormActions.repoError(RepoErrorType.USERNAME_EMPTY)),
    );

    const iteration = getReposIterator.next();
    expect(iteration.done).toBe(true);
  });

  it('should dispatch the reposLoaded action if it requests the data successfully', () => {
    username = 'test';
    repos = [
      {
        name: 'repo1',
        owner: {
          login: 'username1',
        },
      },
    ];

    const requestDescriptor = getReposIterator.next(username).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getReposIterator.next(repos).value;
    expect(putDescriptor).toEqual(
      put(slice.githubRepoFormActions.reposLoaded(repos)),
    );
  });

  it('should dispatch the user not found error', () => {
    username = 'test';

    const requestDescriptor = getReposIterator.next(username).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getReposIterator.throw({
      response: { status: 404 },
    }).value;
    expect(putDescriptor).toEqual(
      put(slice.githubRepoFormActions.repoError(RepoErrorType.USER_NOT_FOUND)),
    );
  });
  it('should dispatch the user has no repo error', () => {
    username = 'test';
    repos = [];

    const requestDescriptor = getReposIterator.next(username).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getReposIterator.next(repos).value;
    expect(putDescriptor).toEqual(
      put(
        slice.githubRepoFormActions.repoError(RepoErrorType.USER_HAS_NO_REPO),
      ),
    );
  });
  it('should dispatch the github rate limit error', () => {
    username = 'test';

    const requestDescriptor = getReposIterator.next(username).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getReposIterator.throw(
      new Error('Failed to fetch'),
    ).value;
    expect(putDescriptor).toEqual(
      put(
        slice.githubRepoFormActions.repoError(RepoErrorType.GITHUB_RATE_LIMIT),
      ),
    );
  });

  it('should dispatch the response error', () => {
    username = 'test';

    const requestDescriptor = getReposIterator.next(username).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getReposIterator.throw(new Error('some error')).value;
    expect(putDescriptor).toEqual(
      put(slice.githubRepoFormActions.repoError(RepoErrorType.RESPONSE_ERROR)),
    );
  });
});

describe('githubRepoFormSaga Saga', () => {
  const githubRepoFormIterator = githubRepoFormSaga();
  it('should start task to watch for loadRepos action', () => {
    const takeLatestDescriptor = githubRepoFormIterator.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(slice.githubRepoFormActions.loadRepos.type, getRepos),
    );
  });
});
