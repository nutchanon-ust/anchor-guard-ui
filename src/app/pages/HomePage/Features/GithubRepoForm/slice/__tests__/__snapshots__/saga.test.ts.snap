// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRepos Saga should dispatch the github rate limit error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRepos Saga should dispatch the github rate limit error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getRepos Saga should dispatch the github rate limit error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://api.github.com/users/test/repos?type=all&sort=updated",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRepos Saga should dispatch the reposLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRepos Saga should dispatch the reposLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getRepos Saga should dispatch the reposLoaded action if it requests the data successfully 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://api.github.com/users/test/repos?type=all&sort=updated",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRepos Saga should dispatch the response error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRepos Saga should dispatch the response error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getRepos Saga should dispatch the response error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://api.github.com/users/test/repos?type=all&sort=updated",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRepos Saga should dispatch the user has no repo error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRepos Saga should dispatch the user has no repo error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getRepos Saga should dispatch the user has no repo error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://api.github.com/users/test/repos?type=all&sort=updated",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRepos Saga should dispatch the user not found error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRepos Saga should dispatch the user not found error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getRepos Saga should dispatch the user not found error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://api.github.com/users/test/repos?type=all&sort=updated",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRepos Saga should return error if username is empty 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRepos Saga should return error if username is empty 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;
